lib/ansible/modules/cloud/amazon/ec2_ami.py

```diff
@@ -428,221 +353,294 @@ def create_image(module, ec2):
+        virtualization_type=dict(default='hvm'),
+        root_device_name=dict(),
+        delete_snapshot=dict(default=False, type='bool'),
+        name=dict(),
+        wait=dict(type='bool', default=False),
+        wait_timeout=dict(default=900, type='int'),
+        description=dict(default=''),
+        no_reboot=dict(default=False, type='bool'),
+        state=dict(default='present'),
+        device_mapping=dict(type='list'),
+        tags=dict(type='dict'),
+        launch_permissions=dict(type='dict'),
+        image_location=dict(),
+        ena_support=dict(type='bool'),
+        billing_products=dict(type='list'),
+        ramdisk_id=dict(),
+        sriov_net_support=dict()
+    ))
+
     module = AnsibleModule(argument_spec=argument_spec)
 
-    if not HAS_BOTO:
-        module.fail_json(msg='boto required for this module')
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 required for this module')
 
     try:
-        ec2 = ec2_connect(module)
-    except Exception as e:
-        module.fail_json(msg="Error while connecting to aws: %s" % str(e))
+        region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+        connection = boto3_conn(module, conn_type='client', resource='ec2', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+        resource = boto3_conn(module, conn_type='resource', resource='ec2', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.NoRegionError:
+        module.fail_json(msg=("Region must be specified as a parameter in AWS_DEFAULT_REGION environment variable or in boto configuration file."))
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Unable to establish connection - " + str(e), exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
 
     if module.params.get('state') == 'absent':
         if not module.params.get('image_id'):
-            module.fail_json(msg='image_id needs to be an ami image to registered/delete')
-
-        deregister_image(module, ec2)
-
+            module.fail_json(msg="The parameter image_id is required to deregister an AMI.")
+        deregister_image(module, connection)
     elif module.params.get('state') == 'present':
         if module.params.get('image_id') and module.params.get('launch_permissions'):
-            # Update image's launch permissions
-            update_image(module, ec2, module.params.get('image_id'))
-
-        # Changed is always set to true when provisioning new AMI
+            update_image(module, connection, module.params.get('image_id'), resource)
         if not module.params.get('instance_id') and not module.params.get('device_mapping'):
-            module.fail_json(msg='instance_id or device_mapping (register from ebs snapshot) parameter is required for new image')
+            module.fail_json(msg="The parameters instance_id or device_mapping (register from EBS snapshot) are required for a new image.")
```

Comment:
Use `required_if`

Reply:
did ['state', 'present', ['name'] but not sure if can do if state=present then x or y (instance_id or device_mapping)?

Reply:
No, you're quite right, you can't do that, as far as I know.